src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1576,63 +1569,101 @@ abstract void onApplicationReply(
 
             private Future<?> signalFuture;
 
-            MqttPublishStream(
+            private int state;
+            private Set<MqttRole> roles;
+
+            MqttServerStream(
                 MessageConsumer application,
                 long routeId,
                 long initialId,
                 long replyId,
                 int packetId,
-                String topicFilter)
+                String topicFilter,
+                Subscription subscription,
+                MqttRole role)
             {
                 this.application = application;
                 this.routeId = routeId;
                 this.initialId = initialId;
                 this.replyId = replyId;
                 this.packetId = packetId;
                 this.topicFilter = topicFilter;
+                this.subscription = subscription;
+                this.roles = EnumSet.of(role);
+                this.subackIndex = subscription != null ? subscription.ackCount : -1;
             }
 
             private void doApplicationBegin(
                 long traceId,
                 long authorization,
-                long affinity)
+                long affinity,
+                String topicFilter,
+                int subscriptionId)
             {
                 assert state == 0;
                 state = MqttState.openingInitial(state);
 
-                this.signalFuture = executor.schedule(publishTimeout, SECONDS, routeId, initialId,
-                                                      PUBLISH_TIMEOUT_SIGNAL);
+                final MqttRole role;
+                if (this.roles.contains(MqttRole.SENDER))
+                {
+                    this.signalFuture = executor.schedule(publishTimeout, SECONDS, routeId, initialId,
```

Comment:
Change `mqtt` server to use `Signaler` instead of `SignalingExecutor` to avoid garbage creation of futures.